packages/hydrojudge/src/task.ts

```diff
@@ -64,8 +98,9 @@ export class JudgeTask {
                 HYDRO_USER: (this.request.uid || 0).toString(),
                 HYDRO_CONTEST: tid,
             };
-            this.next = this.session.getNext(this);
-            this.end = this.session.getEnd(this);
+            this.notifier = new SequentialNotifier();
+            this.next = this.notifier.decorate(this.session.getNext(this));
+            this.end = this.notifier.decorate(this.session.getEnd(this));
```

Comment:
websocket packets are always in order and that should be something need to consider in server-side instead of client-side. if you care about the `record.get` in builtin judge, just cache them in memory (when a new task arrives).

Reply:
In previous implementation, calls to next and end from built-in are not awaited so the order is undetermined. This implementation is to sequentialize these updates to avoid race conditions (also save 100ms latency in end).

Reply:
The `end` function is intentionally delayed to support third-party plugins insert a `next` call after judgement done and before those final results send to user.
For example, a plugin might add an extra message on accept.

Reply:
From my understanding, the built-in judge is not running at the client side and the order of calls to the next / end is not guaranteed so the sequentializer is mainly focused on the built-in judged which running on the server side.

Reply:
I don't want to maintain any code that is only for builtin judge. Either use a standalone one, or rewrite the adapter to connect to `localhost${system.get('server.port')}` and reuse the logic.

Reply:
Connecting to the web socket endpoint requires colossal overhead including network transportation, serialization and deserialization, authentication and file cache logic. I added the sequantializer to maximize the throughout and the simpler implementation would be await for the next function calls when it returns a promise.

Reply:
This builtin adapter is only written to simplify the install logic and not meant to be use under large scale.
Moving to websocket endpoint won't change anything about testdata cache, it will always maintain a cache storage in judge directory (usually `~/.cache/hydro/judge`).
When you have to care about the serialization cost of kilobytes of data, you should definitely move your judge to a seperate machine to spare more resource for judge for better performance and in this case, optimizing builtin judge adapter makes nonsense.

The reason I didn't use websocket for builtin before is that a judge account have to be created when using a judge daemon, but not I got 2 solutions:

1. Grant PRIV_JUDGE to Hydro (UID=1) and use a pre-authed cookie for connection
2. Listen on a extra unix socket which can bypass permission check (maybe not very good)

BTW, the vjudge module will also require a seperate client deployment from next minor version (using the same websocket endpoint as regular judge) and it will be no longer integrated with web server.

Reply:
Okey, I'll revert the change made to the built-in judger for now since it would be deprecated anyway in next minor version and changing it to use web socket is out of the scope of this pull request.